cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java

```diff
@@ -307,16 +311,81 @@ private boolean checkCredIsValid(HttpResponse response) throws OAuthServiceExcep
+    }
+  }
+
+  /** Build HTTP request for getting tokens */
+  private HttpRequest.Builder buildHttpRequest(String body,
+                                               CredentialEncodingStrategy strategy,
+                                               OAuthClientCredentials clientCreds,
+                                               OAuthProvider provider,
+                                               boolean addContentType) throws MalformedURLException {
+    HttpRequest.Builder requestBuilder = HttpRequest.post(new URL(provider.getTokenRefreshURL()))
+            .withBody(body);
+
+    if (addContentType) {
+      requestBuilder.addHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_FORM_URLENCODED);
+    }
+
+    if (strategy == CredentialEncodingStrategy.BASIC_AUTH) {
+      requestBuilder.addHeader(HttpHeaders.AUTHORIZATION, getBasicAuthHeader(clientCreds));
+    }
+
+    return requestBuilder;
+  }
+
+  public HttpRequest createRefreshTokenRequest(CredentialEncodingStrategy strategy,
+                                               String refreshToken,
+                                               OAuthClientCredentials clientCreds,
+                                               OAuthProvider provider) throws MalformedURLException {
+    String body = buildRequestBody(strategy, "refresh_token", null, null, refreshToken, clientCreds);
+    return buildHttpRequest(body, strategy, clientCreds, provider, false).build();
+  }
+
+  public HttpRequest createAccessTokenRequest(CredentialEncodingStrategy strategy,
+                                                    String code,
+                                                    String redirectURI,
+                                                    OAuthClientCredentials clientCreds,
+                                                    OAuthProvider provider) throws MalformedURLException {
+    String body = buildRequestBody(strategy, "authorization_code", code, redirectURI, null, clientCreds);
+    return buildHttpRequest(body, strategy, clientCreds, provider, true).build();
+  }
+
   private HttpRequest createGetRefreshTokenRequest(OAuthProvider provider, String code, String redirectURI)
       throws OAuthServiceException {
     OAuthClientCredentials clientCreds = provider.getClientCredentials();
+    CredentialEncodingStrategy strategy = provider.getCredentialEncodingStrategy();
+
     try {
-      return HttpRequest.post(new URL(provider.getTokenRefreshURL()))
-          .withBody(String.format(
-              "code=%s&redirect_uri=%s&client_id=%s&client_secret=%s&grant_type=authorization_code",
-              code, redirectURI, clientCreds.getClientId(), clientCreds.getClientSecret()))
-          .addHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_FORM_URLENCODED)
-          .build();
+      return createRefreshTokenRequest(strategy, code, redirectURI, clientCreds, provider);
```

Comment:
This should have been `createAccessTokenRequest`

Reply:
I don't understand, I believe these are both correct here

Reply:
createRefreshTokenHttpRequest creates request with grant_type=refresh_token while createAcessTokenHttpRequest creates request with grant_type=refresh_token. From what I can earlier we were creating the request with grant_type=access_token so it should call the latter method here.